api/breaker/breaker_test.go

```diff
@@ -0,0 +1,550 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package breaker
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCircuitBreaker_generation(t *testing.T) {
+	t.Parallel()
+	clock := clockwork.NewFakeClock()
+
+	cb, err := New(Config{
+		Clock:    clock,
+		Interval: time.Second,
+		Trip:     StaticTripper(false),
+	})
+	require.NoError(t, err)
+
+	generation, state := cb.currentState(clock.Now().UTC())
+	require.Equal(t, uint64(1), generation)
+	require.Equal(t, StateStandby, state)
+	require.Equal(t, clock.Now().UTC().Add(time.Second), cb.expiry)
+
+	clock.Advance(500 * time.Millisecond)
+	generation, state = cb.currentState(clock.Now().UTC())
+	require.Equal(t, uint64(1), generation)
+	require.Equal(t, StateStandby, state)
+	clock.Advance(501 * time.Millisecond)
+	generation, state = cb.currentState(clock.Now().UTC())
+	require.Equal(t, uint64(2), generation)
+	require.Equal(t, StateStandby, state)
+	require.Equal(t, clock.Now().UTC().Add(time.Second), cb.expiry)
+
+	for i := 0; i < 1000; i++ {
+		prevGeneration, prevState := cb.currentState(clock.Now().UTC())
+		cb.nextGeneration(clock.Now().UTC())
+		generation, state := cb.currentState(clock.Now().UTC())
+		require.NotEqual(t, prevGeneration, generation)
+		require.Equal(t, prevState, state)
+	}
+
+	generation, state = cb.currentState(clock.Now().UTC())
+	require.Equal(t, uint64(1002), generation)
+	require.Equal(t, StateStandby, state)
+}
+
+func TestCircuitBreaker_beforeRequest(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		desc       string
+		generation uint64
+		executions uint32
+		advance    time.Duration
+		state      State
+		errorCheck require.ErrorAssertionFunc
+	}{
+		{
+			desc:       "standby allows execution",
+			generation: 1,
+			executions: 1,
+			state:      StateStandby,
+			errorCheck: require.NoError,
+		},
+		{
+			desc:       "tripped prevents executions",
+			generation: 1,
+			executions: 0,
+			state:      StateTripped,
+			errorCheck: func(t require.TestingT, err error, i ...interface{}) {
+				require.Error(t, err)
+				require.ErrorIs(t, ErrStateTripped, err)
+			},
+		},
+		{
+			desc:       "recovering before rc allows prevents executions",
+			generation: 1,
+			executions: 0,
+			state:      StateRecovering,
+			errorCheck: func(t require.TestingT, err error, i ...interface{}) {
+				require.Error(t, err)
+				require.ErrorIs(t, ErrRecoveryLimitExceeded, err)
+			},
+		},
+		{
+			desc:       "recovering after rc allows allows executions",
+			generation: 1,
+			executions: 1,
+			state:      StateRecovering,
+			advance:    3 * time.Second,
+			errorCheck: require.NoError,
+		},
+	}
+
+	for _, tt := range cases {
+		tt := tt
+		t.Run(tt.desc, func(t *testing.T) {
+			t.Parallel()
```

Comment:
It might not be worth it to parallelize this small handful of tests individually; either way, we should either mark the test or the subtests as parallel, not both. This applies to all the tests with subtests in `breaker_test.go`.